src/security/cert_generators/file/certificate_db.cc

```diff
@@ -59,13 +59,20 @@ void Ssl::Lock::lock()
         throw TextException(ToSBuf("Failed to open file ", filename), Here());
 
 #if _SQUID_WINDOWS_
-    if (!LockFile(hFile, 0, 0, 1, 0))
+    if (!LockFile(hFile, 0, 0, 1, 0)) {
+        CloseHandle(hFile);
+        hFile = INVALID_HANDLE_VALUE;
 #elif _SQUID_SOLARIS_
-    if (lockf(fd, F_LOCK, 0) != 0)
+    if (lockf(fd, F_LOCK, 0) != 0) {
+        xclose(fd);
+        fd = -1;
 #else
-    if (flock(fd, LOCK_EX) != 0)
+    if (flock(fd, LOCK_EX) != 0) {
+        xclose(fd);
+        fd = -1;
```

Comment:
I agree that this code is buggy, but, AFAICT, this code does not "leak" (as implied by this PR title) because the file descriptor/handler is closed via the Lock class destructor. The real bug is actually _worse_ -- Lock::locked() returns true even though the object is not locked after an flock() failure!

There are several ways to address this problem without duplicating unlock() code inside lock() as currently proposed in this PR. The best way is probably ... to completely delete Ssl::Lock class and use File (from base/File.h) instead. IIRC, it was not possible to reuse File when this code was written, but it should be possible now.

@MegaManSec, would you be willing to try doing that in a dedicated PR while keeping this PR focused on the OPENSSL_free() bug fix?

Reply:
In fact I cannot do this; a mix of lack of time and lack of time to learn. I apologize. Please treat this as a bug report and close.